### PR TITLE
Tca 34 implement logic for login error

### PR DESCRIPTION
**Pull request by pscacchi**
The route from error login fragment to login fragment was implemented.
The action is trigger when the user tap on the error login fragment.

### CODE REVIEW

File: app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt [hunk 1]
```diff
@@ -29,7 +29,7 @@ class FragmentAuthWebDialog : Fragment() {
 
         binding = FragmentAuthWebDialogLayoutBinding.inflate(inflater)
 
-        viewModel.userAccessToken.observe(viewLifecycleOwner) {
+        viewModel.addTokenObserver(viewLifecycleOwner) {
             when {
                 it.error == Constants.ERROR_CANCELLED_AUTH -> {
                     findNavController().navigateUp()
```
Comment (Diana-Endava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt:35]: to validate if the token is "" better to use a function like it.isBlank or isEmpty, choose whatever fits your needs

Comment (Diana-Endava) [reply app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt:35]: Also remember to remove the logs before pushing

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModel.kt [hunk 2]
```diff
@@ -34,7 +37,8 @@ class AuthWebDialogViewModel (
             }
             queryParameters["error"]?.let { error ->
                 if (error == "access_denied") {
```
[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModel.kt:39]: It's was replaced by a constant.

File: app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt (hunk 1 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt:35]: "" was replaced and the Log were deleted.

File: app/src/main/java/ar/scacchipa/twittercloneapp/domain/ErrorLookupCreatorUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package ar.scacchipa.twittercloneapp.domain
+
+import ar.scacchipa.twittercloneapp.datasource.UserAccessToken
+import ar.scacchipa.twittercloneapp.datasource.provideAuthSourceDateApi
+import ar.scacchipa.twittercloneapp.datasource.provideRetrofit
+import ar.scacchipa.twittercloneapp.repository.AuthorizationRepository
+import ar.scacchipa.twittercloneapp.repository.IAuthorizationRepository
+
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/domain/ErrorLookupCreatorUseCase.kt:8]: I dont think this use case is needed.

File: app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt [hunk 4]
```diff
@@ -30,10 +30,15 @@ class AuthorizationRepository(
...
             response.body()?:UserAccessToken()
         } else {
-            UserAccessToken()
+            UserAccessToken(error = Constants.ERROR_HOST_LOOKUP_TOKEN)
         }
     }
-    override fun getErrorUserCaseTokenCreator() : UserAccessToken {
-        return UserAccessToken(error = Constants.ERROR_CODE)
+
+    override fun getCancelledAuthToken() : UserAccessToken {
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt:37]: I dont think this needs to reach the repository.

File: app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt [hunk 5]
```diff
@@ -30,10 +30,15 @@ class AuthorizationRepository(
...
+        return UserAccessToken(error = Constants.ERROR_CANCELLED_AUTH)
+    }
+
+    override fun getErrorLookupToken(): UserAccessToken {
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt:41]: Idem above.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package ar.scacchipa.twittercloneapp.viewmodel
+
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+
+class LoginViewModel: ViewModel() {
+
+    private val _mustShowErrorMsg = MutableLiveData(false)
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt:9]: remember the pattern, private mutableLivedata, public liveData.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    fun showErrorMsg() {
+        _mustShowErrorMsg.value = true
+    }
+
+    fun onClickMsgButton() {
+        _mustShowErrorMsg.value = false
+    }
+
+    fun addMsgChangeObserver( owner: LifecycleOwner, observer: (Boolean) -> Unit) {
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt:20]: I dont think is useful, it brings more problems when testing than solutions.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt [hunk 8]
```diff
@@ -10,17 +10,17 @@ import kotlinx.coroutines.launch
...
         viewModelScope.launch {
-            splashWasSpent.value = splashTimer.spendSplash()
+            _splashWasSpent.value = splashTimer.spendSplash()
         }
     }
     fun getSplashWasSpent(): Boolean {
-        return splashWasSpent.value?:false
+        return _splashWasSpent.value?:false
     }
     fun addObserver(owner: LifecycleOwner, observer: (Boolean) -> Unit) {
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt:23]: same here, not needed.

File: app/src/test/java/ar/scacchipa/twittercloneapp/domain/AuthorizationUseCaseTest.kt [hunk 9]
```diff
@@ -48,7 +48,11 @@ class MockAuthorizationRespository: IAuthorizationRepository {
         )
     }
 
-    override fun getErrorUserCaseTokenCreator(): UserAccessToken {
-        return UserAccessToken(error = "error")
+    override fun getCancelledAuthToken(): UserAccessToken {
+        return UserAccessToken(error = "cancelled_authorization")
+    }
+
+    override fun getErrorLookupToken(): UserAccessToken {
```
Comment (sebastianMullerEndava) [inline app/src/test/java/ar/scacchipa/twittercloneapp/domain/AuthorizationUseCaseTest.kt:55]: same as before, I think it should not reach the repository this functionality.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt [hunk 10]
```diff
@@ -10,17 +10,17 @@ import kotlinx.coroutines.launch
 class SplashViewModel(
```
Comment (Diana-Endava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt:10]: why the changes in the splash? this was already closed. Also we miss the details that you were using mutable instead of liveData

File: app/src/main/java/ar/scacchipa/twittercloneapp/domain/ErrorLookupCreatorUseCase.kt (hunk 3 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/domain/ErrorLookupCreatorUseCase.kt:8]: This class was deleted.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt (hunk 6 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt:9]: It was replaced by:
    private val _mustShowErrorMsg = MutableLiveData(false)
    val mustShowErrorMsg = _mustShowErrorMsg as LiveData<Boolean>

    private val _navToFragAuthWeb = MutableLiveData(false)
    val navTpFragAuthWeb = _navToFragAuthWeb as LiveData<Boolean>

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt (hunk 7 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt:20]: It is unnecessary and It was removed

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    fun onNavToAuthWeb() {
+        _navToFragAuthWeb.value = false
+    }
+
+    fun onClickLoginButton() {
+        _navToFragAuthWeb.value = true
+    }
+
+    fun addNavigatorObserver(owner: LifecycleOwner, observer: (Boolean) -> Unit) {
```
[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginViewModel.kt:32]: Item too.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt [hunk 12]
```diff
@@ -10,17 +10,17 @@ import kotlinx.coroutines.launch
...
     private val splashTimer: SplashTimerUseCase = SplashTimerUseCase()
 ): ViewModel() {
-    private val splashWasSpent = MutableLiveData(false)
```
[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt:13]: It was replace by:
    private val _splashWasSpent = MutableLiveData(false)
    val splashWasSpent = _splashWasSpent as LiveData<Boolean>

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt (hunk 8 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt:23]: Idem above

File: app/src/test/java/ar/scacchipa/twittercloneapp/domain/AuthorizationUseCaseTest.kt (hunk 9 above)
Comment (pscacchi) [reply app/src/test/java/ar/scacchipa/twittercloneapp/domain/AuthorizationUseCaseTest.kt:55]: I was removed and the object is created in the class.

File: app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt (hunk 10 above)
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/SplashViewModel.kt:10]: Because I didn't want to explose a MutableLiveData. It was fixed._

File: app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt [hunk 13]
```diff
@@ -28,13 +29,21 @@ class FragmentAuthWebDialog : Fragment() {
...
 
         viewModel.userAccessToken.observe(viewLifecycleOwner) {
-            if (it.accessToken != "") {
-                val action = FragmentAuthWebDialogDirections
-                    .actionFragmentAuthWebDialogToFragmentHome(it.accessToken)
-                findNavController().navigate(action)
-            } else {
-                findNavController().navigateUp()
-            }
+            when {
```
[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt:32]: It wes replaced by: viewModel.userAccessToken.observe(viewLifecycleOwner) { token ->
            when {
                token.isCancelationErrorToken() -> {
                    findNavController().navigateUp()
                }
                token.isHostLookupErrorToken()-> {
                    val action= FragmentAuthWebDialogDirections
                        .actionFragmentLoginAuthWebDialogToFragmentLogin(true)
                    findNavController().navigate(action)
                }
                token.isSuccessToken() -> {
                    val action = FragmentAuthWebDialogDirections
                        .actionFragmentAuthWebDialogToFragmentHome(token.accessToken)
                    findNavController().navigate(action)
                }
           }
        }

and 
In the class UserAccessToken, it's was added:
data class UserAccessToken (
    @SerializedName("token_type") val tokenType: String = "",
    @SerializedName("expires_in") val expiresIn: Int = 0,
    @SerializedName("access_token") val accessToken: String = "",
    @SerializedName("scope") val scope: String = "",
    @SerializedName("refresh_token") val refreshToken: String = "",
    @SerializedName("") val error: String = "",
    @SerializedName("error_description") val errorDescription: String = ""
) {
    fun isCancelationErrorToken(): Boolean {
        return error == Constants.ERROR_CANCELLED_AUTH
    }
    fun isHostLookupErrorToken(): Boolean {
        return error == Constants.ERROR_HOST_LOOKUP_TOKEN
    }
    fun isSuccessToken(): Boolean {
        return accessToken.isNotBlank()
    }
}

File: app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt [hunk 14]
```diff
@@ -30,10 +30,15 @@ class AuthorizationRepository(
         if (response.isSuccessful) {
```
[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt:30]: I'm working in this issue.

[the remark this answers is not in the document]
Comment (pscacchi) [reply app/src/main/java/ar/scacchipa/twittercloneapp/repository/AuthorizationRepository.kt]: It was replace by sealed class

File: app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt [hunk 15]
```diff
@@ -27,13 +28,25 @@ class FragmentAuthWebDialog : Fragment() {
...
         binding = FragmentAuthWebDialogLayoutBinding.inflate(inflater)
 
-        viewModel.userAccessToken.observe(viewLifecycleOwner) {
-            if (it.accessToken != "") {
-                val action = FragmentAuthWebDialogDirections
-                    .actionFragmentAuthWebDialogToFragmentHome(it.accessToken)
-                findNavController().navigate(action)
-            } else {
-                findNavController().navigateUp()
+        viewModel.tokenResource.observe(viewLifecycleOwner) { token ->
```
Comment (sebastianMullerEndava) [inline app/src/main/java/ar/scacchipa/twittercloneapp/fragment/FragmentAuthWebDialog.kt:31]: nice :)

File: app/src/test/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModelTest.kt [hunk 16]
```diff
@@ -49,56 +49,53 @@ class AuthWebDialogViewModelTest {
                 "scope=tweet.read%20tweet.write" +
                 "state=state&code_challenge=challenge&code_challenge_method=plain")
 
-        subject.controlRequest(uri)
-        Assert.assertEquals(subject.userAccessToken.value, null)
+
+        Assert.assertEquals(subject.tokenResource.value, null)
```
Comment (sebastianMullerEndava) [inline app/src/test/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModelTest.kt:53]: I think you can delete this test.

**Review by sebastianMullerEndava — APPROVED** [review]

**Review by Diana-Endava — APPROVED** [review]

File: app/src/test/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModelTest.kt (hunk 16 above)
Comment (pscacchi) [reply app/src/test/java/ar/scacchipa/twittercloneapp/viewmodel/AuthWebDialogViewModelTest.kt:53]: It will be removed for the next PR (about test).
